classifier.py

```diff
@@ -3130,59 +3090,436 @@ def target_age(cls, identifier, name):
+            if target_min is not None:
+                if not c.subject.target_age.lower_inc:
+                    target_min += 1
+                for i in range(0,c.weight):
+                    target_age_mins.append(target_min)
+            if target_max is not None:
+                if not c.subject.target_age.upper_inc:
+                    target_max -= 1
+                for i in range(0,c.weight):
+                    target_age_maxes.append(target_max)
+
+        target_age_min = None
+        target_age_max = None
+        if target_age_mins:
+            # We have opinions about the proper minimum age.
+            mins = Counter(target_age_mins)
+            [(target_age_min, count)] = mins.most_common(1)
+            if count == 1:
+                # Everyone has a different opinion. Pick the smallest
+                # minimum age.
+                target_age_min = min(target_age_mins)
+
+        if target_age_maxes:
+            # We have opinions about the proper maximum age.
+            maxes = Counter(target_age_maxes)
+            [(target_age_max, count)] = maxes.most_common(1)
+            if count == 1:
+                # Everyone has a different opinion. Pick the largest
+                # maximum age.
+                target_age_max = max(target_age_maxes)
+
+        if not target_age_min and not target_age_max:
+            # We found no opinions about target age. Use the default.
+            return Classifier.default_target_age_for_audience(audience)
+
+        if target_age_min is None:
+            target_age_min = target_age_max
+
+        if target_age_max is None:
+            target_age_max = target_age_min
+
+        # If min and max got mixed up somehow, un-mix them.
+        if target_age_min > target_age_max:
+            target_age_min, target_age_max = target_age_max, target_age_min
+        return target_age_min, target_age_max
+
+    def genres(self, fiction, cutoff=0.15):
+        """Consolidate genres and apply a low-pass filter."""
+        # Remove any genres whose fiction status is inconsistent with the
+        # (independently determined) fiction status of the book.
+        #
+        # It doesn't matter if a book is classified as 'science
+        # fiction' 100 times; if we know it's nonfiction, it can't be
+        # science fiction. (It's probably a history of science fiction
+        # or something.)
+        genres = dict(self.genre_weights)
+        for genre in list(genres.keys()):
+            if genre.default_fiction != fiction:
+                del genres[genre]
+
+        # Consolidate parent genres into their heaviest subgenre.
+        genres = self.consolidate_genre_weights(genres)
+        total_weight = float(sum(genres.values()))
+
+        # Strip out the stragglers.
+        for g, score in list(genres.items()):
+            affinity = score / total_weight
+            if affinity < cutoff:
+                total_weight -= score
```

Comment:
I think the idea is that as the big genres are removed from the pool, the smaller genres have a better chance of meeting the cutoff. I'm not convinced that this is a good idea, but here's why I think I came up with it. To take an extreme example let's say it's:

Science Fiction 100000 (99%)
Romance 1000 (0.99%)
Western 1 (0.00099%)
Total: 101001

Clearly the primary classification here is Science Fiction. Western is obviously out. But is it fair to exclude Romance, which has a really high score, just because SF has an even higher score? I think my idea was to recalibrate what "high score" meant after each round.

In retrospect I think this idea needs work, at the very least. But my inclination is to leave it alone for now and have Courteney take a look at it.